runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/QueryablePipeline.java

```diff
@@ -61,12 +66,23 @@
    * <p>The returned {@link QueryablePipeline} will contain only the primitive transforms present
    * within the provided components.
    */
-  public static QueryablePipeline fromComponents(Components components) {
+  public static QueryablePipeline forPrimitivesIn(Components components) {
+    return forComponents(retainOnlyPrimitives(components));
+  }
+
+  /**
+   * Create a new {@link QueryablePipeline} based on the provided components.
+   *
+   * <p>Relationships between composite transforms and their subtransforms, and producer
+   * relationships between {@link PTransformNode transforms} and {@link PCollectionNode
+   * PCollections} are not yet modelled by {@link QueryablePipeline}, so the input {@link
+   * Components} should be treatable as though each node is a primitive.
+   */
+  static QueryablePipeline forComponents(Components components) {
```

Comment:
It would make sense to make ExecutableStage a primitive and have its payload be the `subtransforms`, `coders`, `pcollections`, ...

This will address some of the coder issues and the duality where you want the runner based pipeline representation to differ from the SDK based pipeline segment which will now be completely embedded inside the executable stage.

Reply:
https://github.com/apache/beam/pull/4844 performs a lot of this change, without embedding the components within the stage.